server/src/main/java/org/elasticsearch/gateway/GatewayService.java

```diff
@@ -197,117 +212,202 @@ public void clusterChanged(final ClusterChangedEvent event) {
+                        indicesService.verifyIndexMetadata(metaData, metaData);
+                    }
+                } catch (final Exception e) {
+                    final Index electedIndex = metaData.getIndex();
+                    logger.warn(() -> new ParameterizedMessage("recovering index {} failed - recovering as closed", electedIndex), e);
+                    metaData = IndexMetaData.builder(metaData).state(IndexMetaData.State.CLOSE).build();
                 }
+                builder.put(metaData, false);
+            }
 
+            return ClusterState.builder(clusterState).metaData(builder).build();
+        }
+
+        static ClusterState updateRoutingTable(final ClusterState state) {
+            // initialize all index routing tables as empty
+            final RoutingTable.Builder routingTableBuilder = RoutingTable.builder(state.routingTable());
+            for (final ObjectCursor<IndexMetaData> cursor : state.metaData().indices().values()) {
+                routingTableBuilder.addAsRecovery(cursor.value);
+            }
+            // start with 0 based versions for routing table
+            routingTableBuilder.version(0);
+            return ClusterState.builder(state).routingTable(routingTableBuilder.build()).build();
+        }
+
+        static ClusterState removeStateNotRecoveredBlock(final ClusterState state) {
+            return ClusterState.builder(state)
+                    .blocks(ClusterBlocks.builder()
+                            .blocks(state.blocks()).removeGlobalBlock(STATE_NOT_RECOVERED_BLOCK).build())
+                    .build();
+        }
+
+        static ClusterState mixCurrentStateAndRecoveredState(final ClusterState currentState, final ClusterState recoveredState) {
+            assert currentState.metaData().indices().isEmpty();
+
+            final ClusterBlocks.Builder blocks = ClusterBlocks.builder()
+                    .blocks(currentState.blocks())
+                    .blocks(recoveredState.blocks());
+
+            final MetaData.Builder metaDataBuilder = MetaData.builder(recoveredState.metaData());
+            // automatically generate a UID for the metadata if we need to
+            metaDataBuilder.generateClusterUuidIfNeeded();
+
+            for (final IndexMetaData indexMetaData : recoveredState.metaData()) {
+                metaDataBuilder.put(indexMetaData, false);
+            }
+
+            return ClusterState.builder(currentState)
+                    .blocks(blocks)
+                    .metaData(metaDataBuilder)
+                    .build();
+        }
+
+    }
+
+    class RecoverStateUpdateTask extends ClusterStateUpdateTask {
+
+        @Override
+        public ClusterState execute(final ClusterState currentState) {
+            final ClusterState newState = Function.<ClusterState>identity()
+                    .andThen(state -> Updaters.closeBadIndices(state, indicesService))
```

Comment:
this now moves the execution of this (possibly computationally expensive) code to the cluster state update task. I'm not sure yet what kind of impact that has, and wonder whether we should do more of this for Zen2 during startup instead of a post-election step. The advantage of running this process at startup is that we can fail the node if anything goes wrong here, instead of just logging. The current solution also has the disadvantage that it has potentially already published a state that is unrecoverable / cannot upgrade / archive invalid settings, possibly poisoning other nodes.
Can you see what it would take to move this (`Updaters.closeBadIndices` + `Updaters.upgradeAndArchiveUnknownOrInvalidSettings`) to the place where we start (currently the place called `GatewayMetaState.setLocalNode`). We can pass IndicesService as constructor argument to GatewayMetaState in the Node class so have all deps available to run these static methods in GatewayMetaState's `setLocalNode` method (which obviously will need to be renamed).

Reply:
Sure. I've done the following changes:
1. Made `ClusterStateUpdaters` top-level class. Refactored out tests for it in `ClusterStateUpdatersTests` file.
2. Replaced `GatewayMetaState.setLocalNode` with `GatewayMetaState.applyClusterStateUpdaters`. It accepts no arguments, I retrieve `localNode` from `transportService`, which is now passed as a constructor argument.
3. `applyClusterStateUpdaters` now does as much work as possible, also I've moved `closedBadIndices` and `upgradeAndArchiveUnknownOrInvalidSettings` for Zen1 back to Gateway.

https://github.com/elastic/elasticsearch/pull/36013/commits/b1e0aa8153432fe014f3192738ed5c5e3332de1f